File: bench/sumtype.c
#include <stdio.h>
typedef enum { North, East, South, West } Dir;
static Dir turn(Dir d){ switch(d){ case North:return East; case East:return South; case South:return West; default:return North; } }
static long val(Dir d){ switch(d){ case North:return 0; case East:return 1; case South:return 2; default:return 3; } }
static Dir from_int(long n){ switch(n%4){ case 0:return North; case 1:return East; case 2:return South; default:return West; } }
static long inner(Dir d, long acc, long n){ while(n){ acc=(acc+val(d))%1000000L; d=turn(d); n--; } return acc; }
int main(void){ long acc=0; for(long k=4000;k;k--) acc=(acc+inner(from_int(k),0,50000))%2147483647L; printf("%ld\n", acc); return 0; }

File: bench/sumtype.rs
#[derive(Clone, Copy)]
enum Dir { North, East, South, West }
fn turn(d: Dir) -> Dir { match d { Dir::North=>Dir::East, Dir::East=>Dir::South, Dir::South=>Dir::West, Dir::West=>Dir::North } }
fn val(d: Dir) -> i64 { match d { Dir::North=>0, Dir::East=>1, Dir::South=>2, Dir::West=>3 } }
fn from_int(n: i64) -> Dir { match n % 4 { 0=>Dir::North, 1=>Dir::East, 2=>Dir::South, _=>Dir::West } }
fn inner(mut d: Dir, mut acc: i64, mut n: i64) -> i64 { while n != 0 { acc = (acc + val(d)) % 1000000; d = turn(d); n -= 1; } acc }
fn main() { let mut acc: i64 = 0; let mut k: i64 = 4000; while k != 0 { acc = (acc + inner(from_int(k), 0, 50000)) % 2147483647; k -= 1; } println!("{}", acc); }
